File: codewars/authoring/dart/isomorphism.dart
/// so, when are two type, `a` and `b`, considered equal?
/// a definition might be, it is possible to go from `a` to `b`,
/// and from `b` to `a`.
/// Going a roundway trip should leave you the same value.
/// Unfortunately it is virtually impossible to test this in Haskell.
/// This is called ISO.
class ISO<A, B> {
  Function _bw, _fw;
  ISO(this._fw, this._bw);
  /// go backward
  A bw(final B b) => _bw(b);
  /// go forward
  B fw(final A a) => _fw(a);
}

abstract class Either<L, R> {
  T pm<T>(Function lt, Function rt);
  dynamic get();
  static _Left<L, R> left<L, R>(L l) => new _Left(l);
  static _Right<L, R> right<L, R>(R r) => new _Right(r);
  static isLeft(Either e) => e is _Left;
  static isRight(Either e) => e is _Right;
}

class _Left<L, R> extends Either<L, R> {
  L _l;
  _Left(this._l);
  @override dynamic get() => _l;
  @override T pm<T>(Function lt, Function rt) => lt(_l);
  @override bool operator ==(Object rhs) => (rhs as Either<L, R>).pm((o) => _l == o, (rr) => false);
}

class _Right<L, R> extends Either<L, R> {
  L _r;
  _Right(this._r);
  @override dynamic get() => _r;
  @override T pm<T>(Function lt, Function rt) => rt(_r);
  @override bool operator ==(Object rhs) => (rhs as Either<L, R>).pm((ll) => false, (o) => _r == o);
}

class Tuple<A, B> {
  A a;
  B b;
  Tuple(this.a, this.b);
  @override bool operator ==(other) => other is Tuple<A, B> && other.a == a && other.b == b;
}

abstract class Optional<T> {
  static _None _none = new _None();
  static _Some<T> from<T> (T obj) => new _Some(obj);
  static _None empty() => _none;
  @override bool operator ==(other);
  Optional<T> map(Function f);
  Optional<T> fmap(Function f);
  T get();
  T orElseGet(Function f);
  bool isPresent();
}

class _Some<T> extends Optional<T> {
  T _obj;
  _Some(this._obj);
  @override _Some<T> map(Function f) => new _Some(f(_obj));
  @override Optional<T> fmap(Function f) => f(_obj);
  @override bool operator ==(other) => other is _Some && _obj == other._obj;
  @override T get() => _obj;
  @override T orElseGet(Function f) => _obj;
  @override bool isPresent() => true;
}

class _None<T> extends Optional<T> {
  _None();
  @override _None map(Function f) => this;
  @override _None fmap(Function f) => this;
  @override bool operator ==(other) => true;
  /// don't change this
  @override T get() => throw "why";
  @override T orElseGet(Function f) => f();
  @override bool isPresent() => false;
}

abstract class Void {
  A absurd<A>();
}

/// given ISO a b, we can go from a to b
Function subStL<A, B>(ISO<A, B> iso) => iso.fw;

/// and vise versa
Function subStR<A, B>(ISO<A, B> iso) => iso.bw;

/// There can be more than one ISO a b
ISO<bool, bool> isoBool() => refl();

ISO<bool, bool> isoBoolNot() => new ISO((a) => !a, (b) => !b);

/// isomorphism is reflexive
ISO<A, A> refl<A>() => new ISO((a) => a, (b) => b);

/// isomorphism is symmetric
ISO<A, B> symm<A, B>(final ISO<B, A> iso) => new ISO(iso.bw, iso.fw);

/// isomorphism is transitive
ISO<A, C> trans<A, B, C>(final ISO<A, B> ab, final ISO<B, C> bc) => new ISO((a) => bc.fw(ab.fw(a)), (c) => ab.bw(bc.bw(c)));

/// we can combine isomorphism
ISO<Tuple<A, C>, Tuple<B, D>> isoTuple<A, B, C, D>(final ISO<A, B> ab, final ISO<C, D> cd) => new ISO((ac) => new Tuple(ab.fw(ac.a), cd.fw(ac.b)), (bd) => new Tuple(ab.bw(bd.a), cd.bw(bd.b)));

ISO<List<A>, List<B>> isoList<A, B>(final ISO<A, B> iso) => new ISO((as) => as.map(iso.fw), (bs) => bs.map(iso.bw));

ISO<Optional<A>, Optional<B>> isoOptional<A, B>(final ISO<A, B> iso) => new ISO((a) => a.map(iso.fw), (b) => b.map(iso.bw));

ISO<Either<A, C>, Either<B, D>> isoEither<A, B, C, D>(final ISO<A, B> ab, final ISO<C, D> cd) =>
    new ISO((Either l) => l.pm((a) => Either.left(ab.fw(a)), (c) => Either.right(cd.fw(c))), (Either r) => r.pm((b) => Either.left(ab.bw(b)), (d) => Either.right(cd.bw(d))));

ISO<Either<List<Null>, Null>, Either<List<Null>, Void>> isoEU() =>
    new ISO((Either le) => le.pm((l) => Either.left(l..insert(0, l)), (r) => Either.left([])),
            (Either re) => re.pm((l) => l.isNotEmpty ? Either.left(l..removeAt(0)) : Either.right(null), (r) => r.absurd()));

/// Going another way is hard (and is generally impossible)
/// Remember, for all valid ISO, converting and converting back
/// is the same as the original value.
/// You need this to prove some case are impossible.
ISO<A, B> isoUnOptional<A, B>(final ISO<Optional<A>, Optional<B>> iso) =>
    new ISO((a) => iso.fw(Optional.from(a)).orElseGet(() => iso.fw(Optional.empty()).get()), (b) => iso.bw(Optional.from(b)).orElseGet(() => iso.bw(Optional.empty()).get()));

/// hint: Function<A, C>, Function<B, D>
ISO<Function, Function> isoFunc<A, B, C, D>(final ISO<A, B> ab, final ISO<C, D> cd) =>
    new ISO((ac) => (a) => cd.fw(ac(ab.bw(a))), (bd) => (b) => cd.bw(bd(ab.fw(b))));

/// And we have isomorphism on isomorphism!
ISO<ISO<A, B>, ISO<B, A>> isoSymm<A, B>() => new ISO(symm, symm);

File: codewars/authoring/cs/Isomorphism/Isomorphism.cs
using System;
using System.Collections.Generic;
using System.Linq;

// ReSharper disable InconsistentNaming
#pragma warning disable 693
#pragma warning disable 659

public abstract class Either<L, R>
{
	public abstract T Match<T>(Func<L, T> lt, Func<R, T> rt);
	public abstract override bool Equals(object obj);
	public static _Left<L, R> Left(L l)
	{
		return new _Left<L, R>(l);
	}

	public static _Right<L, R> Right(R r)
	{
		return new _Right<L, R>(r);
	}

	public class _Left<L, R> : Either<L, R>
	{
		private readonly L l;

		public _Left(L l)
		{
			this.l = l;
		}

		public override T Match<T>(Func<L, T> lt, Func<R, T> rt)
		{
			return lt(l);
		}

		public override bool Equals(object rhs)
		{
			return rhs is Either<L, R>
			       && ((Either<L, R>) rhs).Match(arg => l.Equals(arg), rr => false);
		}
	}

	public class _Right<L, R> : Either<L, R>
	{
		private readonly R r;

		public _Right(R r)
		{
			this.r = r;
		}

		public override T Match<T>(Func<L, T> lt, Func<R, T> rt)
		{
			return rt(r);
		}

		public override bool Equals(object rhs)
		{
			return rhs is Either<L, R>
			       && ((Either<L, R>) rhs).Match(ll => false, arg => r.Equals(arg));
		}
	}
}

public abstract class Void
{
	public abstract A Absurd<A>();
}

public class Unit
{
	private Unit()
	{
	}

	public static readonly Unit INSTANCE = new Unit();
}

public abstract class Optional<T>
{
	public static Some<T> From(T obj)
	{
		return new Some<T>(obj);
	}

	public static None<T> Empty()
	{
		return new None<T>();
	}

	public abstract Optional<R> Map<R>(Func<T, R> f);
	public abstract Optional<R> fmap<R>(Func<T, Optional<R>> f);
	public abstract T Get();
	public abstract T OrElseGet(Func<T> f);
	public abstract bool IsPresent();
	public abstract override bool Equals(object obj);

	public class Some<T> : Optional<T>
	{
		private readonly T _obj;

		public Some(T obj)
		{
			_obj = obj;
		}

		public override Optional<R> Map<R>(Func<T, R> f)
		{
			return new Some<R>(f(_obj));
		}

		public override Optional<R> fmap<R>(Func<T, Optional<R>> f)
		{
			return f(_obj);
		}

		public override bool Equals(object other)
		{
			return other is Some<T> && Equals(_obj, ((Some<T>) other)._obj);
		}

		public override T Get()
		{
			return _obj;
		}

		public override T OrElseGet(Func<T> f)
		{
			return _obj;
		}

		public override bool IsPresent()
		{
			return true;
		}
	}

	public class None<T> : Optional<T>
	{
		public None()
		{
		}

		public override Optional<R> Map<R>(Func<T, R> f)
		{
			return new None<R>();
		}

		public override Optional<R> fmap<R>(Func<T, Optional<R>> f)
		{
			return new None<R>();
		}

		/// don't change this
		public override T Get()
		{
			throw new Exception("cannot get From null");
		}

		public override bool Equals(object obj)
		{
			return obj is None<T>;
		}

		public override T OrElseGet(Func<T> f)
		{
			return f();
		}

		public override bool IsPresent()
		{
			return false;
		}
	}
}

public class ISO<A, B>
{
	public Func<A, B> Fw;
	public Func<B, A> Bw;

	public ISO(Func<A, B> fw, Func<B, A> bw)
	{
		Fw = fw;
		Bw = bw;
	}
}

/// so, when are two type, `a` and `b`, considered equal?
/// a definition might be, it is possible to go from `a` to `b`,
/// and from `b` to `a`.
/// Going a roundway trip should leave you the same value.
/// Unfortunately it is virtually impossible to test this in Haskell.
/// This is called Isomorphism.
///
/// ISO can also be simulated in C#.
///
/// Created by ice1000 on 17-6-19.
///
public static class Isomorphism
{
	/// given ISO a b, we can go from a to b
	public static Func<A, B> SubStL<A, B>(ISO<A, B> iso)
	{
		return iso.Fw;
	}

	/// and vise versa
	public static Func<B, A> SubStR<A, B>(ISO<A, B> iso)
	{
		return iso.Bw;
	}

	/// isomorphism is reflexive
	public static ISO<A, A> Refl<A>()
	{
		return new ISO<A, A>(a => a, a => a);
	}

	public static ISO<bool, bool> IsoBool()
	{
		return Refl<bool>();
	}

	public static ISO<bool, bool> IsoBoolNot()
	{
		return new ISO<bool, bool>(a => !a, b => !b);
	}

	/// isomorphism is symmetric
	public static ISO<A, B> Symm<A, B>(ISO<B, A> iso)
	{
		return new ISO<A, B>(iso.Bw, iso.Fw);
	}

	/// isomorphism is transitive
	public static ISO<A, C> Trans<A, B, C>(ISO<A, B> ab, ISO<B, C> bc)
	{
		return new ISO<A, C>(a => bc.Fw(ab.Fw(a)), c => ab.Bw(bc.Bw(c)));
	}

	/// we can combine isomorphism
	public static ISO<Tuple<A, C>, Tuple<B, D>> IsoTuple<A, B, C, D>(ISO<A, B> ab, ISO<C, D> cd)
	{
		return new ISO<Tuple<A, C>, Tuple<B, D>>(
			ac => Tuple.Create(ab.Fw(ac.Item1), cd.Fw(ac.Item2)),
			bd => Tuple.Create(ab.Bw(bd.Item1), cd.Bw(bd.Item2)));
	}

//	public static ISO<Either<L, R>, Either<R, L>> IsoEither<L, R>() =>
//		new ISO<Either<L, R>, Either<R, L>>(
//			lr => lr.Match<Either<R, L>>(Either<R, L>.Right, Either<R, L>.Left),
//			rl => rl.Match<Either<L, R>>(Either<L, R>.Right, Either<L, R>.Left));

	public static ISO<Either<A, C>, Either<B, D>> IsoEither<A, B, C, D>(ISO<A, B> ab, ISO<C, D> cd)
	{
		return new ISO<Either<A, C>, Either<B, D>>(
			l => l.Match<Either<B, D>>(a => Either<B, D>.Left(ab.Fw(a)), c => Either<B, D>.Right(cd.Fw(c))),
			r => r.Match<Either<A, C>>(b => Either<A, C>.Left(ab.Bw(b)), d => Either<A, C>.Right(cd.Bw(d))));
	}

	public static ISO<List<A>, List<B>> IsoList<A, B>(ISO<A, B> iso)
	{
		return new ISO<List<A>, List<B>>(a => a.ConvertAll(i => iso.Fw(i)), b => b.ConvertAll(i => iso.Bw(i)));
	}

	public static ISO<Optional<A>, Optional<B>> IsoOptional<A, B>(ISO<A, B> iso)
	{
		return new ISO<Optional<A>, Optional<B>>(a => a.Map(iso.Fw), b => b.Map(iso.Bw));
	}

	// We cannot have
	// isoUnEither ::
	// ISO (Either a b) (Either c d) -> ISO a c -> ISO b d.
	// Note that we have
	public static ISO<Either<List<Unit>, Unit>, Either<List<Unit>, Void>> IsoEU()
	{
		return new ISO<Either<List<Unit>, Unit>, Either<List<Unit>, Void>>(
			le => le.Match<Either<List<Unit>, Void>>(l =>
				{
					l.Insert(0, Unit.INSTANCE);
					return Either<List<Unit>, Void>.Left(l);
				},
				r => Either<List<Unit>, Void>.Left(new List<Unit>())),
			re => re.Match(l => l.Count != 0
				? (Either<List<Unit>, Unit>) Either<List<Unit>, Unit>.Left(l.Skip(1).ToList())
				: Either<List<Unit>, Unit>.Right(Unit.INSTANCE), r => r.Absurd<Either<List<Unit>, Unit>>()));
	}
	// where (), the empty tuple, has 1 value, and Void has 0 value
	// If we have isoUnEither,
	// We have ISO () Void by calling isoUnEither isoEU
	// That is impossible, since we can get a Void by
	// substL on ISO () Void
	// So it is impossible to have isoUnEither


	/// Going another way is hard (and is generally impossible)
	/// Remember, for all valid ISO, converting and converting back
	/// is the same as the original value.
	/// You need this to prove some case are impossible.
	public static ISO<A, B> IsoUnOptional<A, B>(ISO<Optional<A>, Optional<B>> iso)
	{
		return new ISO<A, B>(
			a => iso.Fw(Optional<A>.From(a)).OrElseGet(() => iso.Fw(Optional<A>.Empty()).Get()),
			b => iso.Bw(Optional<B>.From(b)).OrElseGet(() => iso.Bw(Optional<B>.Empty()).Get()));
	}

	public static ISO<Func<A, C>, Func<B, D>> IsoFunc<A, B, C, D>(ISO<A, B> ab, ISO<C, D> cd)
	{
		return new ISO<Func<A, C>, Func<B, D>>(ac => a => cd.Fw(ac(ab.Bw(a))), bd => b => cd.Bw(bd(ab.Fw(b))));
	}

	/// And we have isomorphism on isomorphism!
	public static ISO<ISO<A, B>, ISO<B, A>> IsoSymm<A, B>()
	{
		return new ISO<ISO<A, B>, ISO<B, A>>(Symm, Symm);
	}
}